### PR TITLE
Miscellaneous cleanup

### PR DESCRIPTION
**Pull request by Ralith**
Straightforward improvements extracted from #1578.

### CODE REVIEW

File: quinn-proto/src/endpoint.rs [hunk 1]
```diff
@@ -748,17 +691,115 @@ impl fmt::Debug for Endpoint {
...
             .field("connections", &self.connections)
             .field("config", &self.config)
             .field("server_config", &self.server_config)
             .finish()
     }
 }
 
+/// Maps packets to existing connections
+#[derive(Default, Debug)]
+struct ConnectionIndex {
```
Comment (djc) [inline quinn-proto/src/endpoint.rs:706]: Not sure `Index` feels like the most obvious suffix here. Maybe `Map`?

File: quinn-proto/src/endpoint.rs [hunk 2]
```diff
@@ -748,17 +691,115 @@ impl fmt::Debug for Endpoint {
...
+            })
+            .or_else(|| {
+                let data = datagram.data();
+                if data.len() < RESET_TOKEN_SIZE {
+                    return None;
+                }
+                self.connection_reset_tokens
+                    .get(addresses.remote, &data[data.len() - RESET_TOKEN_SIZE..])
+            })
+            .cloned()
```
Comment (djc) [inline quinn-proto/src/endpoint.rs:796]: In this unnested context, I think this could benefit from rewriting with early returns (and the `?` operator) and explicit matches rather than this combinator festival...

File: quinn-proto/src/endpoint.rs [hunk 3]
```diff
@@ -559,15 +548,18 @@ impl Endpoint {
...
                 trace!(id = ch.0, icid = %dst_cid, "connection incoming");
-                Some((ch, conn))
+                Some(DatagramEvent::NewConnection(ch, conn))
             }
             Err(e) => {
                 debug!("handshake failed: {}", e);
                 self.handle_event(ch, EndpointEvent(EndpointEventInner::Drained));
                 if let ConnectionError::TransportError(e) = e {
-                    self.initial_close(version, addresses, crypto, &src_cid, &loc_cid, e);
+                    Some(DatagramEvent::Response(self.initial_close(
```
Comment (djc) [inline quinn-proto/src/endpoint.rs]: Style nit: maybe replace the `if let` with a `match` here?

File: quinn-proto/src/endpoint.rs [hunk 4]
```diff
@@ -129,7 +121,7 @@ impl Endpoint {
         local_ip: Option<IpAddr>,
         ecn: Option<EcnCodepoint>,
         data: BytesMut,
-    ) -> Option<(ConnectionHandle, DatagramEvent)> {
+    ) -> Option<DatagramEvent> {
```
Comment (djc) [inline quinn-proto/src/endpoint.rs:124]: I think this could be factored out into a separate commit? Seems like a somewhat independent change, would make it a little easier to review.

File: quinn-proto/src/tests/util.rs [hunk 5]
```diff
@@ -310,26 +310,25 @@ impl TestEndpoint {
...
                             .push_back(event);
                     }
+                    DatagramEvent::Response(transmit) => {
+                        self.outbound.extend(split_transmit(transmit));
+                    }
                 }
             }
         }
 
-        while let Some(x) = self.poll_transmit() {
```
Comment (djc) [inline quinn-proto/src/tests/util.rs:329]: I suppose this changes the order of things. Have you reasoned about why it's okay to change things? Wasn't there a reason to do "metadata" events before I/O?

File: quinn-proto/src/endpoint.rs [hunk 6]
```diff
@@ -62,7 +61,6 @@ impl Endpoint {
     ) -> Self {
         Self {
             rng: StdRng::from_entropy(),
-            transmits: VecDeque::new(),
```
Comment (djc) [inline quinn-proto/src/endpoint.rs:81]: Certainly a nice direction!

File: quinn-proto/src/endpoint.rs [hunk 7]
```diff
@@ -575,7 +582,8 @@ impl Endpoint {
         tls: Box<dyn crypto::Session>,
         server_config: Option<Arc<ServerConfig>>,
         transport_config: Arc<TransportConfig>,
-    ) -> (ConnectionHandle, Connection) {
+        ch: ConnectionHandle,
```
Comment (djc) [inline quinn-proto/src/endpoint.rs:585]: This feels out of order with respect to our usual style?

**Review by djc — COMMENTED** [review]
Nice work! Should we proactively bump the quinn-proto version incompatibly?

File: quinn-proto/src/endpoint.rs (hunk 1 above)
Comment (Ralith) [reply quinn-proto/src/endpoint.rs:706]: I think `Index` is descriptive here. A map owns its values; an index (as in database index) merely refers to them.

File: quinn-proto/src/endpoint.rs (hunk 2 above)
Comment (Ralith) [reply quinn-proto/src/endpoint.rs]: Good call. We can't use `?` here because the desired fallback is always "try the next case", but it's still a fair bit more readable this way.

File: quinn-proto/src/endpoint.rs (hunk 3 above)
Comment (Ralith) [reply quinn-proto/src/endpoint.rs]: Done.

File: quinn-proto/src/endpoint.rs (hunk 4 above)
Comment (Ralith) [reply quinn-proto/src/endpoint.rs:124]: Done.

File: quinn-proto/src/endpoint.rs (hunk 7 above)
Comment (Ralith) [reply quinn-proto/src/endpoint.rs]: Fixed.

**Comment by Ralith** [comment]
> Should we proactively bump the quinn-proto version incompatibly?

Ah, right, forgot the new policy.

File: quinn-proto/src/tests/util.rs (hunk 5 above)
Comment (Ralith) [reply quinn-proto/src/tests/util.rs:329]: I don't think this changes any behavior, because appending to `self.outbound` has no side effects.

Comment (djc) [reply quinn-proto/src/tests/util.rs:329]: Ah, fair. So what about for other downstream users of the quinn-proto crate?

**Review by djc — APPROVED** [review]

File: quinn-proto/src/tests/util.rs (hunk 5 above)
Comment (Ralith) [reply quinn-proto/src/tests/util.rs:329]: Transmits not associated with any connection are by definition not associated with any persistent state, so no special care should be required, even leaving aside the fact that we could never rely on the ordering of real I/O to begin with.
